twister: cmake: Add prefiltration based on cmake package helper script

PerMac opened the pull request:
Twister allows filtering based on kconfigs and dts, however the filtration is a part of the cmake stage, i.e. the stage has to pass first and then twister checks if required properties are available. This causes problems, when the full cmake stage is unable to pass. If so, other filtration methods had to be used, e.g. platform_allow. The commit modifies the twister workflow:
if a test defines filters based on kconfig/dts first a cmake package helper script is used to extract dt and kconfigs and if the conditions are fulfilled  it proceeds to a regular cmake stage. If not, test is skipped.

Co-authored-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
Signed-off-by: Maciej Perkowski Maciej.Perkowski@nordicsemi.no

PerMac committed:
twister: cmake: Add prefiltration based on cmake package helper script

Twister allows filtering based on kconfigs and dts, however the
filtration is a part of the cmake stage, i.e. the stage has to pass
first and then twister checks if required properties are available.
This causes problems, when the full cmake stage is unable to pass.
If so, other filtration methods had to be used, e.g. platform_allow.
The commit modifies the twister workflow:
if a test defines filters based on kconfig/dts first a cmake package
helper script is used to extract dt and/or kconfigs and if the
conditions are fulfilled  it proceeds to a regular cmake stage.
If not, test is skipped.

Co-authored-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -274,6 +274,83 @@ def run_build(self, args=[]):
 
+    def run_filter_cmake(self, args=""):
+
+        if not self.options.disable_warnings_as_errors:
```
You are duplicating alot of code from run_cmake. Please figure out how to define and do this only in one place

PerMac replied:
@nashif I reworked the PR removing duplicated methods and reusing the existing ones. It ended up with only a small expansion needed

nashif reviewed (changes requested):
still looking into that, but first thing that I noticed is the duplicated code, we need to find a better way to do this and reuse the existing code instead of duplicating it.

PerMac commented:
This PR should enhance viability of using dt/kconfig filters. Before, such filtration was often not possible as the full cmake would fail. This is e.g. needed for the mcuboot tests https://github.com/zephyrproject-rtos/zephyr/pull/49552#discussion_r964896155

gopiotr commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -296,6 +296,17 @@ def run_cmake(self, args=""):
+        # testcases that should be built
+        if filter:
+            cmake_filter_args = [
+                f'-S{self.source_dir}',
```
Is this necessary to add here `-S{self.source_dir}`? If I understand it correctly, the source is passed to CMake command few lines later.

PerMac commented:
@nashif Changes in the workflow and example of usage:

if "filter:" is NOT USED in testcase/sample.yaml: no changes to the workflow

if "filter:" is USED in testcase/sample.yaml start the workflow from an extra stage: `"filter()"`. This stage calls the existing cmake stage, but expands with extra args used by package helper:
```
        if filter:
            filter_stages = self.get_cmake_filter_stages(instance)
        res = self.run_cmake(args_expanded, filter_stages)
```
and
```
        if filter_stages:
            cmake_filter_args = [
                f'-DMODULES={",".join(filter_stages)}',
                f'-P{canonical_zephyr_base}/cmake/package_helper.cmake',
            ]
``` 

Filter stages can contain "dts" and/or "kconfig" keys. It tells which stages are requried from cmake package helper to perform the filtration. 

Then the package helper script is called to extract dts and/or kconfig without completing the whole cmake (https://github.com/zephyrproject-rtos/zephyr/pull/41302). 
Next, twister checks if the test should be skipped by matching the entries from "filter:" from yaml with extracted dts and kconfig. Those are the same checks which were already used for the same purpose, however currently those checks are run after the full cmake. The point is that there are many cases, where we couldn't filter based on dts and/or kconfig because the full cmake was failing and had to use methods like platform_allow.

The test for mcuboot with sysbuild is an example of such issue. @danieldegrasse had to use platform_allow to limit the execution. However, there are more platforms that can run the tests. According to @de-nordic the platforms able to run mcuboot can be distinguished by:
```
filter: dt_nodelabel_enabled("boot_partition") and dt_nodelabel_enabled("slot0_partition") 
            and dt_nodelabel_enabled("slot1_partition")
``` 
This shows the issue with the current workflow: Such filter cannot be used since platforms without mcuboot configured would fail the full cmake and we won't be able to do the filtration. One can replace platform_allow in testcase.yaml with the above filter and try executing 
`scripts/twister -v --platform=nrf52840dk_nrf52840 --platform=nucleo_l073rz --platform=nsim_sem -T tests/boot/test_mcuboot`

Without the PR:
- nrf52840dk_nrf52840: the filter finds matching nodes and build continues. Good.
- nucleo_l073rz: twister fails with cmake dt build failure:
``` 
devicetree error: /chosen: undefined node label 'boot_partition'

```
- nsim_sem: fails with cmake kconfig error:
```
warning: FLASH_MAP (defined at subsys/storage/flash_map/Kconfig:10) has direct dependencies FLASH_HAS_DRIVER_ENABLED with value n, but is currently being y-selected by the following symbols:
 - MCUBOOT_IMG_MANAGER (defined at subsys/dfu/Kconfig:25), with value y, direct dependencies <choice> (value: y), and select condition <choice> (value: y)

warning: FLASH_PAGE_LAYOUT (defined at drivers/flash/Kconfig:54) has direct dependencies FLASH_HAS_PAGE_LAYOUT && FLASH with value n, but is currently being y-selected by the following symbols:
 - STREAM_FLASH (defined at subsys/storage/stream/Kconfig:7), with value y, direct dependencies y (value: y)

warning: USE_DT_CODE_PARTITION (defined at Kconfig.zephyr:88) has direct dependencies HAS_FLASH_LOAD_OFFSET with value n, but is currently being y-selected by the following symbols:
 - BOOTLOADER_MCUBOOT (defined at Kconfig.zephyr:762), with value y, direct dependencies !MCUBOOT (value: y), and select condition !MCUBOOT (value: y)

error: Aborting due to Kconfig warnings

```

With this PR (since with it only dt is required for filtering):
```
INFO    - 1/3 nsim_sem                  zephyr/tests/boot/test_mcuboot/boot.mcuboot        SKIPPED (runtime filter)
INFO    - 2/3 nucleo_l073rz             zephyr/tests/boot/test_mcuboot/boot.mcuboot        SKIPPED (runtime filter)
INFO    - 3/3 nrf52840dk_nrf52840       zephyr/tests/boot/test_mcuboot/boot.mcuboot        PASSED (build)
```

I discussed the usage of package helper with @tejlmand and adapted the workflow with his suggestions. Torsten, can you back up on why this change is useful?

@danieldegrasse I had to modify the PR after your acceptance. I had to add stages selection for package helper script, so we can quit just after dt in case kconfigs are not required for filtration. It was needed for mcuboot as the example above shows that some platforms will fail the kconfig cmake stage.

PerMac commented:
I had to edit the above description. The workflow is changed only when there is "filter:" in the yaml, AND only "dt_" or "CONFIG" entries are used there. This change is needed, since there are other keys (in particular TOOLCHAIN_HAS_NEWLIB) which require cmake cache to work. Package helper doesn't produce this cache. In such case, the regular workflow is performed, where filters are checked after full cmake

hakehuang commented:
can we update the twister document for this feature? also add a test for this?

danieldegrasse commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -1038,3 +1081,33 @@ def execute(self, pipeline, done):
+            expressions = [expression for expression in expressions if expression not in expr_parser.reserved.keys()]
+            for expression in expressions:
+                # all dt based expressions have the same prefix. Hence, expressions without it relate to kconfig.
+                if "dt_" in expression:
```
Shouldn't these checks only be verifying that the expression starts with `dt_` or `CONFIG_`? Testing this in a python3 REPL:
```
>>> expression="hello_dt_world"
>>> "dt_" in expression
True
```
Given this, some variable like CMAKE_CACHE_CONFIG_VAR might fall into the `elif "CONFIG" in expression` block. I'm not sure if we have any filter variables that would cause this, just something to consider.

tejlmand commented:
> I discussed the usage of package helper with @tejlmand and adapted the workflow with his suggestions. Torsten, can you back up on why this change is useful?

I think the comment by @PerMac https://github.com/zephyrproject-rtos/zephyr/pull/53499#issuecomment-1400100993 summarizes one of the advantages very well.
Developers don't need to just add a single or two platforms to `platform_allow:` when what they really want is to have a given test cases runable  on platform meeting the requirement.
Problem today is that because of Kconfig failure, and thus CMake failure, then it must manually be decided what platform to choose.

Another advantage is performance. When having a lot of boards, then it's faster to only run code until devicetree has been processed and then do filtering, instead of running whole CMake, including Kconfig, compiler setup, CMake code for libraries, etc. for platforms that will anyway be skipped.

I did some measurement long time back where running dts was ~170ms vs full CMake ~890ms, (number exclude the CMake startup of ~350ms, don't ask me what CMake is doing in those 350ms).

One may say it's a small gain, but running sufficient number of times, then it adds up.
And on top of that, we can have more test suites capable of running on boards, as they no longer need to be limited by `platform_allow`, as the dt filter will be working properly.

nashif commented:
@PerMac  can you please address some of the feedback above so we can move forward with this?

PerMac commented:
@nashif Done.
@gopiotr the change was indeed necessary, fixed. 
@hakehuang I expanded a bit a section about "filter" keyword. Also added unit tests for my method deciding which cmake stages are required.
@danieldegrasse I modified the method and now an expression string hast to starts with the given prefix.  
The best solution would be to use expr_parser.py module to determine if dt/kconfig/full filtration is needed, however since https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/expr_parser.py is a black box to me I don't even know if it is possible.

nashif commented:
needs a rebase.

evgeniy-paltsev commented:
@PerMac I see massive regression when building / running tests - and git besect points to this commit.

The simphtoms are that we get Cmake erroe while building, i.e:

```
$ cat twister-out/nsim_hs/tests/kernel/common/kernel.common.tls/build.log
CMake Error at zephyr/cmake/package_helper.cmake:53 (function):
  Built-in flow control command "add_custom_target" cannot be overridden.
```

The error occurs in ~100 tests per twister run.

Component versions:
```
Zephyr SDK: 0.15.0
west_version: 0.14.0 
make_version: 4.3 
ninja_version: 1.10.2 
cmake_version: 3.20.2 
```

Any thoughts?

tejlmand commented:
initial question, is package_helper invoked in configure and not script mode ?

tejlmand commented:
tested locally with different CMake versions:
**Fails with**: CMak 3.20.2
**Works with**: 3.21.1, 3.22.6, 3.23.1, 3.25.1

so seems to be a CMake regression on specific CMake versions.

tejlmand commented:
As suspected, it's a CMake regression:
https://gitlab.kitware.com/cmake/cmake/-/issues/22310

and fixed here:
https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6232

PerMac commented:
Just for the record: Upstream CI is using docker which has cmake 3.20.5 and it works as well. Right now zephyr requires 3.20.0 as min CMake.

tejlmand commented:
Proposed fix: https://github.com/zephyrproject-rtos/zephyr/pull/55547

PerMac commented:
@evgeniy-paltsev Can you switch to higher cmake>=3.20.5? I won't comment on which cmake has to be the minimal required since it is not my domain. If needed I can propose a patch to twister, which will check version of cmake and if < 3.20.5 to ignore  calling cmake package helper script and do the old workflow, where only full cmake is called with filtration afterwards. Also if too low cmake is detected a warning will be raised explaining that package helper is disable without needed cmake version. @nashif what is your suggestion? Should I provide such patch?

evgeniy-paltsev commented:
> Can you switch to higher cmake>=3.20.5?

Will try, thanks!
